### ISSUE TITLE
[FSDP2] HSDP with globally sharded fp32 weights and optimizer states

### ISSUE THREAD

**Issue by ChrisLiu6** [pytorch/pytorch#148257]
### 🚀 The feature, motivation and pitch

First, I hope to show respect to the FSDP/FSDP2 team. I have been using FSDP for a long time and am recently working on migrating to FSDP2. I feel the new APIs are now more user-friendly and logically sound. Thanks so much for your efforts!

I am writing to inquiry if the following feature can/will be supported by FSDP2:
Let's consider HSDP as the starting point. When working with HSDP + gradient accumulation, I want 
1. (not supported) the FP32 model parameters and optimizer states are sharded globally across the whole DP_replicate * DP_shard ranks. 
2. (supported) After forward, the parameters are resharded onto the intra-node 8 gpus (which can be achieved with `reshard_after_forward=True`
3. (supported) After backward, the parameters and gradients are resharded onto the intra-node 8 gpus **if not the last gradient accumulation step**
    By the way, I currently implement this through the following, is that correct?
```
y = model(x)
model.set_requires_gradient_sync(True, recurse=True)
model.set_requires_all_reduce(last_accumulate_step, recurse=True)
model.set_reshard_after_backward(True, recurse=True)
model.set_reshard_after_backward(last_accumulate_step, recurse=False)
y.backward()
```
4. (unsupported) at the last accumulation step, I hope the intra-node-sharded gradients to be reduced-scattered, instead of all-reduced, among all dp_replicate\*dp_shard ranks. Parameters are optimized on each rank, and afterwards in the next first forward, the updated parameters are all-gathered from dp_replicate\*dp_shard ranks

I also think the aforementioned is equivalent to, considering 1d mesh (namely not HSDP) as the starting point, the following (suppose we have 4 nodes * 8GPUs/node = 32 GPUs:
1. (supported) after forward, shard the parameters with each node, which supported by `reshard_after_forward=8`
2. (not supported) after backward, if not last accumulation step, parameters are resharded **within each node** instead of globally (semanticallly similar to `reshard_after_backlward=8, which is unsupported as only bool values are accepted`
3. (not supported) similar to parameters, the gradients should also be sharded with node if not last accumulation step
4. after the last accumulation step, reduce_scatter the accumulated and intra-node-sharded gradients among the total 32 gpus to update

I think this functionality can help FSPD2 better handle situations where large gradient accumulation is needed. Don't hesitate to correct me if I make anything wrong. Looking forward to your reply!

Thanks!

@awgu

**Comment by weifengpy** [comment]
thanks for the proposal! this looks like an interesting trade off. I want to understand more for prioritization.

comparing to plain HSDP, are you proposing this to save more memory by sharing optimizers globally?

comparing to plain FSDP, are you proposing this for QPS gain? I feel QPS gain could be offset by resharding op after forward and backward. wonder if you are confident abou the gain?

**Comment by weifengpy** [comment]
@mori360 another case of global sharded optimizers for HSDP. But this proposal is about fully shard in the last microbatch of  HSDP2 forward/backward

**Comment by weifengpy** [comment]
another Q: if we can do standard HSDP for fwd/bwd, and do `optimizer.unshard` -> `optimizer.unshard` -> `optimizer.reshard`, would it achieve your goal as well?

**Comment by ChrisLiu6** [comment]
> thanks for the proposal! this looks like an interesting trade off. I want to understand more for prioritization.
> 
> comparing to plain HSDP, are you proposing this to save more memory by sharing optimizers globally?
> 
> comparing to plain FSDP, are you proposing this for QPS gain? 

Thanks for reply! Yes, this is what I am expecting from the proposal.


> I feel QPS gain could be offset by resharding op after forward and backward. wonder if you are confident abou the gain?

I think the comfortable zone of this proposed strattegy is when holding the total N parameters in memory is too costly for one GPU but holding N/8 is acceptable (*so resharding after forward and backward can not be avoided anyway*). Currently in such cases we have to use ZERO3-like FSDP (which induces costly cross-node all-gather, whereas the proposed can change it to more efficient intra-node all-gather), or HSDP, which  costs a lot more memory as the huge [(2 adam states + 1 fp32 weight + 1 ema weight) * N parameters * 4Bytes/param] memory cost is distributed on less GPUs. 
Besides, even in the case where one gpu can afford the unsharded full N parameters in bf16 (so reshard_after_forward/backward can be disabled), the gradient can still be a problem: if graidents are also kept unshareded before the last micro batch (similar to zero1), then since 16-bit gradient accumulation is unstable, we need to spend N * 4bytes/param for the accumulated gradients; in contrast, if we reduce scatter the gradients globally after each microbatch (similar to zero2), then this means costly cross-node reduce-scatter after each microbatch (however, as the reduce scatter can be overlapped with backward computation, I don't know if the actual delay would be as obvious as I estimate)


> another Q: if we can do standard HSDP for fwd/bwd, and do optimizer.unshard -> optimizer.unshard -> optimizer.reshard, would it achieve your goal as well?

I am not quite sure what ` optimizer.unshard -> optimizer.unshard -> optimizer.reshard` exactly means, but I guess if we merely operate on the optimizer, then the FP32 model parameters are still not globally sharded? Besides, when EMA model is needed, this may also mean that the ema weights cannot be easilly globally sharded

**Comment by weifengpy** [comment]
thanks for the explaining the motivation in detail. I got your point now. We want to make FSDP2 resharding as flexible as possible. will think more about the complexity to achieve this
